mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -244,51 +187,55 @@ struct MagicStructCode {
-                             ty: &syn::Ty,
-                             idx: usize) -> (quote::Tokens, quote::Tokens) {
-    let slot_type_name = quote::Ident::from(format!("{}SlotIndex{}", name, idx));
-    let slot = quote! {
-        enum #slot_type_name {}
-        impl SlotIndex for #slot_type_name {
-            fn slot_index() -> u32 { #idx as u32 }
-        }
-    };
-    let field = quote! {
-        #id: MagicSlot<#ty, #slot_type_name>
-    };
-    (field, slot)
-}
-
 fn gen_getter(id: &Option<syn::Ident>,
               name: &quote::Ident,
-              ty: &syn::Ty) -> (quote::Tokens, quote::Ident) {
+              idx: u32) -> (quote::Tokens, quote::Ident) {
     let getter_str = match *id {
         Some(ref real_id) => format!("get_{}", real_id.to_string()),
         None => panic!("Encounter a empty field. Something wrong..."),
     };
     let getter_name = quote::Ident::from(getter_str);
     let getter = quote! {
-        pub unsafe fn #getter_name (obj: &#name, cx: *mut jsapi::JSContext) -> #ty {
-            (*obj).#id.get(cx)
+        pub unsafe fn #getter_name<T: ToJSValConvertible + FromJSValConvertible<Config = ()>> (
+            obj: &#name, cx: *mut jsapi::JSContext) -> T {
+            let jsobj = obj.object;
+            rooted!(in(cx) let val = jsapi::JS_GetReservedSlot(jsobj, #idx));
+
+            let conversion = FromJSValConvertible::from_jsval(cx, val.handle(), ())
+                .expect("Should never put anything into a MagicSlot that we can't \
+                         convert back out again");
+
+            match conversion {
+                ConversionResult::Success(v) => v,
+                ConversionResult::Failure(why) => {
+                    panic!("Should never put anything into a MagicSlot that we \
+                            can't convert back out again: {}",
+                           why);
+                }
+            }
         }
     };
     (getter, getter_name)
 }
 
 fn gen_setter(id: &Option<syn::Ident>,
               name: &quote::Ident,
-              ty: &syn::Ty) -> (quote::Tokens, quote::Ident) {
+              idx: u32) -> (quote::Tokens, quote::Ident) {
     let setter_str = match *id {
         Some(ref real_id) => format!("set_{}", real_id.to_string()),
         None => panic!("Encounter a empty field. Something wrong..."),
     };
     let setter_name = quote::Ident::from(setter_str);
     let setter = quote! {
-        pub fn #setter_name (obj: &#name, cx: *mut jsapi::JSContext, t: #ty) {
+        pub fn #setter_name<T: ToJSValConvertible + FromJSValConvertible<Config = ()>> (
+            obj: &#name,
+            cx: *mut jsapi::JSContext,
+            t: T) {
             unsafe {
-                (*obj).#id.set(cx, t);
+                let jsobj = obj.object;
+                rooted!(in(cx) let mut val = jsval::UndefinedValue());
+                t.to_jsval(cx, val.handle_mut());
+                jsapi::JS_SetReservedSlot(jsobj, #idx, &*val);
```

Comment:
Again, we don't want to store any arbitrary `T` that can be converted to and from a JS value, we only want to store values of the expected type here.